File: 3.Longest_Substring_Without_Repeating_Characters/python/main.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = 0
        maxLength = 0
        charIndex = {}

        for right, char in enumerate(s):
            if char in charIndex.keys():
                if charIndex[char]>=left:
                    left = charIndex[char]+1
 
            charIndex[char] = right
            currentLength = right - left + 1

            maxLength = max(currentLength, maxLength)
            

        return maxLength

File: 3.Longest_Substring_Without_Repeating_Characters/golang/main.go
package main

import "fmt"

func main() {
	fmt.Println(lengthOfLongestSubstring("abcabcbb"))
}

func lengthOfLongestSubstring(s string) int {
	leftBoundary := 0
	maxLength := 0
	charIndex := make(map[rune]int)

	for rightBoundary, char := range s {
		lastIndex, ok := charIndex[char]
		if ok && lastIndex >= leftBoundary {
			leftBoundary = lastIndex + 1
		}

		charIndex[char] = rightBoundary

		currentLength := rightBoundary - leftBoundary + 1

		if currentLength > maxLength {
			maxLength = currentLength
		}

	}

	return maxLength

}
